### PR TITLE
first chapter of ALF tutorial

### PR DESCRIPTION
**Pull request by hnyu**
This is the first chapter of an ALF tutorial in Issue #934 . The outline will adjust as the tutorial proceeds. Please do provide comments and suggestions, or even contribute to the tutorial.

### CODE REVIEW

File: docs/tutorial.rst [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+separate source code files with little code reuse, ALF categorizes RL algorithms
+and distills the common structure and logic within each caterogy, so
+that each algorithm only needs to implement or override its own exclusive logic. The
+categorization and distillation processes require huge effort, as a deep understanding
+of different algorithms is needed to know the relations and differences among
+algorithms.
+
+However, once correctly done, modularization provides at least two benefits:
+
+1. *Fast prototyping without worring about breaking things.*
```
Comment (breakds) [inline docs/tutorial.rst]: Nit: s/worring/worrying

File: docs/tutorial.rst [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+which implement different algorithms by repeating the almost entire RL pipeline in
+separate source code files with little code reuse, ALF categorizes RL algorithms
+and distills the common structure and logic within each caterogy, so
+that each algorithm only needs to implement or override its own exclusive logic. The
+categorization and distillation processes require huge effort, as a deep understanding
+of different algorithms is needed to know the relations and differences among
+algorithms.
+
+However, once correctly done, modularization provides at least two benefits:
+
```
Comment (breakds) [inline docs/tutorial.rst:11]: It might be good to explain some high level terminologies here. (I might have named them wrong though) Based on my limited experience with alf, in general to use alf:

- The user is expected to write **configuration** to create an alf project (or a better terminology).
- In each alf project the user needs to configure both the **environment** (which is task specific) and the **algorithm**.
- With such configuration, the alf project can be executed to perform training, evaluation and more.

The benefits paragraph follows naturally.

File: docs/tutorial.rst [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+     is used,
+   * A convenient episodic memory infrastructure for easily implementing
+     models that are stateful,
+   * Automatically applying various input data transformers during rollout
+     and training,
+   * Specifying different optimizers for different sub-algorithms,
+   * Exploiting a variety of tensorboard summary utils,
+   * and many more...
+
+   RL training is a very complex process; some details are very tricky and
```
Comment (breakds) [inline docs/tutorial.rst]: IMHO this paragraph can be promoted to be ahead of the detailed benefits.

File: docs/tutorial/a_minimal_working_example.rst [hunk 4]
```diff
@@ -0,0 +1,128 @@
+A minimal working example
+=========================
+
+We start with a minimal working example of ALF. The example, as a pure ALF
+configuration file, is located at ``<ALF_ROOT>/alf/examples/tutorial/minimal_example_conf.py``,
+and consists of only 8 lines.
+
+Train and play
```
Comment (breakds) [inline docs/tutorial/a_minimal_working_example.rst:8]: I am not sure `play` is a well understood terminology. If not it might worth explaining. For example: play is the procedure of applying the trained agent/policy to perform the task and possibly also visualize it.

File: docs/tutorial/a_minimal_working_example.rst [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+Let's ignore its content for a moment (see the next section
+:doc:`./understanding_ALF_via_the_minimal_working_example` for an explanation of
+the configuration content), and just focus on how to launch the training,
+interpret the output training messages, and play a trained model.
+
+Train from scratch
+^^^^^^^^^^^^^^^^^^
+
+We can train from scratch by
```
Comment (breakds) [inline docs/tutorial/a_minimal_working_example.rst:19]: Do you think it might benefit the first-time users if the concept of `root dir` (a.k.a. state directory, a.k.a. operating direcotry)?

File: docs/tutorial/a_minimal_working_example.rst [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+::
+
+    minimal_example_conf.py -> alf_tutorial1: 0 time=0.046 throughput=5169.30
+
+which has the formatting template as
+
+::
+
+    [conf_file_name] -> [training_root_dir]: [iteration_numer] time=[current_time_per_training_iter] throughput=[current_training_throughput]
```
Comment (breakds) [inline docs/tutorial/a_minimal_working_example.rst:54]: Here we probably want to briefly explain what is a "training iteration"?

File: docs/tutorial/a_minimal_working_example.rst [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    python -m alf.bin.play --root_dir /tmp/alf_tutorial1 --record_file /tmp/alf_tutorial1.mp4
+
+We recommend the reader to read the various commandline flags in ``<ALF_ROOT>/alf/bin/play.py``,
+for specifying different options such as checkpoint number and number of episodes to
+evaluate.
+
+Use Tensorboard for monitoring the training progress
+----------------------------------------------------
+
+While the training is ongoing, we can monitor the real-time progress by
```
Comment (breakds) [inline docs/tutorial/a_minimal_working_example.rst]: IMHO this logically follows the training section rather than play section. What do you think?

File: docs/tutorial/a_minimal_working_example.rst [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+.. code-block::
+
+    tensorboard --logdir /tmp/alf_tutorial1
+
+We leave the interpretation of various Tensorboard statistics to later sections.
+
+Conclusion
```
Comment (breakds) [inline docs/tutorial/a_minimal_working_example.rst]: Looks more like a "Summary" rather than "Conclusion". Maybe rephrase?

**Review by breakds — COMMENTED** [review]
Thanks for the great initiative Haonan!

Left some comments. Just my 2 cents.

File: docs/tutorial/a_minimal_working_example.rst (hunk 6 above)
Comment (hnyu) [reply docs/tutorial/a_minimal_working_example.rst:54]: I'll leave some core concepts to the next section.

**Comment by hnyu** [comment]
@breakds Thanks for your suggestions. They are really helpful. I've updated this PR.

**Review by breakds — APPROVED** [review]
Thanks Haonan!
